### DISCUSSION TITLE
How to add a line number to every line in a manuscript?

### DISCUSSION THREAD

**Discussion by bguo068** [typst/typst#2899] [General]
I am trying to switch from latex to typst. One of the challenges I have is to add a line number (commonly used to prepare manuscripts). In latex, it would look like this:

<img width="538" alt="image" src="https://github.com/typst/typst/assets/13037898/d5b13a8f-6761-42da-acb0-0196caac3cc9">

Is there a similar function/package available in typst? 
Thank you!

**Comment by yochem** [comment]
If you find a way to do this, I would suggest to create a PR for it at https://github.com/ntjess/typst-drafting! I tried to do this, but it didn't work with other elements like figures, headings and tables.

**Reply by bguo068** [reply]
Thank you for the reply. `typst-drafting` seems to be a great place for going deeper into typst. I will see if I can get it and do something fun.

**Comment by yochem** [comment]
I'm currently working on this. It does not work yet with paragraphs that break over pages, but it's a start. I also first tried to do it with `#show par:` rules, but it's just so much easier when defining a custom rule.

```typst
#set text(12pt)

#let marginNumber = counter("marginNumber")

#show pagebreak: it => {
  marginNumber.update(0)
  it
}

= heading

#let placeNumbers(xloc, ystart, yend, sidemargin: 30pt) = {
  style(styles => {
    let stepsize = measure("1", styles).height * 2 - 0.1pt // text height
    let yloc = ystart
    while yloc < yend {
      place(
        top,
        clearance: 0pt,
        dx: xloc - sidemargin,
        dy: yloc,
        marginNumber.display(num => {
          if num < 10 {
            [0#num]
          } else { num }
        })
      )
      yloc += stepsize
      marginNumber.step()
    }
  })
}

#let numbered(body) = [
  #metadata("")<numbered-par-start>
  #body
  #metadata("")<numbered-par-end>

  #let labelPosition(label, loc) = query(
    selector(label).before(loc),
    loc
  ).last().location().position()

  #locate(loc => {
    let parstart = labelPosition(<numbered-par-start>, loc)
    let parend = labelPosition(<numbered-par-end>, loc)

    let xstart = parstart.x - 2.5cm
    let ystart = parstart.y - 2.5cm
    let yend = parend.y - 2.5cm

    placeNumbers(xstart, ystart, yend)
  })
]


#numbered(lorem(300))
#numbered(lorem(200))
```

Result:

![image](https://github.com/typst/typst/assets/23235841/369c0645-95d3-4379-85e1-7adb606b132a)

**Reply by bguo068** [reply]
This already looked amazing to me. It might also be great to have a built in api that toggles on a per line metadata labels. Thank you for the enlightening answer!!

**Comment by jehaj** [comment]
I found this while trying to do the same in my project. Typst allows for this natively now, so the next time I or someone else searches and finds this post, they can just use

![image](https://github.com/user-attachments/assets/3bf99195-24a2-4be9-a279-4868ee2d34b9)

for more information see [Paragraph Line](https://typst.app/docs/reference/model/par/#definitions-line).
